api/base/parsers.py

```diff
@@ -15,30 +15,49 @@ class JSONAPIParser(JSONParser):
     media_type = 'application/vnd.api+json'
     renderer_class = JSONAPIRenderer
 
+    # Overrides JSONParser
     def parse(self, stream, media_type=None, parser_context=None):
         """
-        Parses the incoming bytestream as JSON and returns the resulting data
+        Parses the incoming bytestream as JSON and returns the resulting data.
         """
         result = super(JSONAPIParser, self).parse(stream, media_type=media_type, parser_context=parser_context)
         if not isinstance(result, dict):
             raise ParseError()
         data = result.get('data', {})
 
-        if data:
-            if 'attributes' not in data:
-                raise JSONAPIException(source={'pointer': '/data/attributes'}, detail=NO_ATTRIBUTES_ERROR)
-            id = data.get('id')
-            object_type = data.get('type')
-            attributes = data.get('attributes')
+        def data_flattener(resource_object, stream):
```

Comment:
It may be better to make this a class method (rather than a function in an function). How would we for example unit test this function?